File: master/server/src/playlist.rs
use crate::config::NodeConfig;
use crate::toml_file::ReadAndWriteTOMLFile;
use chrono::prelude::*;
// use serde_derive::Deserialize;

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct Playlist {
    pub working_hours: Option<WorkingHours>,
    pub music: Option<Music>,
    pub advertizement: Option<Advertizement>,
    pub time_announcement: Option<TimeAnnouncement>,
}

#[derive(Deserialize, Serialize, Debug, Clone, Default)]
pub struct WorkingHours {
    pub schedule: Option<Vec<WorkingHoursSchedule>>,
    pub exceptions: Option<HashMap<chrono::NaiveDate, WorkingHoursSchedule>>,
}

#[derive(Deserialize, Serialize, Debug, Clone, Default)]
pub struct WorkingHoursSchedule(
    #[serde(with = "toml_datetime_compat")] pub chrono::NaiveTime,
    #[serde(with = "toml_datetime_compat")] pub chrono::NaiveTime,
);

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct Music {
    pub shuffle: Option<bool>,
    pub schedule: Option<Vec<MusicSchedule>>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct MusicSchedule(
    #[serde(with = "toml_datetime_compat")] pub chrono::NaiveDate,
    #[serde(with = "toml_datetime_compat")] pub chrono::NaiveDate,
    Vec<String>,
);

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct Advertizement {
    pub schedule: Option<HashMap<chrono::NaiveTime, Vec<String>>>,
    pub start_jingle: Option<String>,
    pub end_jingle: Option<String>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct TimeAnnouncement {
    pub folder: Option<String>,
}

impl ReadAndWriteTOMLFile<Playlist> for Playlist {}

File: master/client/src/models/Playlist.ts
export type TOMLDateTime = { $__toml_private_datetime: string }

export type WorkingHoursScheduleElement = [TOMLDateTime, TOMLDateTime]

export type WorkingHoursException = { [d: string]: WorkingHoursScheduleElement }

export type WorkingHoursSchedule = [
  WorkingHoursScheduleElement,
  WorkingHoursScheduleElement,
  WorkingHoursScheduleElement,
  WorkingHoursScheduleElement,
  WorkingHoursScheduleElement,
  WorkingHoursScheduleElement,
  WorkingHoursScheduleElement
]

export type WorkingHours = { exceptions: WorkingHoursException; schedule: WorkingHoursSchedule }

export type MusicScheduleElement = [startDate: TOMLDateTime, endDate: TOMLDateTime, dirs: string[]]

export type Music = { shuffle: boolean; schedule: MusicScheduleElement[] }

export type AdvertizementSchedule = {
  [s: string]: string[]
}

export type Advertizement = {
  start_jingle: string
  end_jingle: string
  schedule?: AdvertizementSchedule
}

export type Playlist = {
  working_hours?: WorkingHours
  music?: Music
  advertizement?: Advertizement
}
